nest-tax-backend/src/tax/utils/unified-tax.util.ts

```diff
@@ -0,0 +1,406 @@
+    throw new ThrowerErrorGuard().InternalServerErrorException(
+      CustomErrorTaxTypesEnum.MISSING_INSTALLMENT_AMOUNTS,
+      CustomErrorTaxTypesResponseEnum.MISSING_INSTALLMENT_AMOUNTS,
+    )
+  }
+
+  let remainingPaid = overallPaid
+  return amounts.map((amount) => {
+    const paid = Math.min(amount, remainingPaid)
+    const toPay = amount - paid
+
+    remainingPaid -= paid
+
+    let status
+    if (toPay === 0) {
+      status = InstallmentPaidStatusEnum.PAID
+    } else if (toPay === amount) {
+      status = InstallmentPaidStatusEnum.NOT_PAID
+    } else {
+      status = InstallmentPaidStatusEnum.PARTIALLY_PAID
+    }
+
+    return { toPay, paid, status }
+  })
+}
+
+const calculateInstallmentPaymentDetails = (
+  overallAmount: number,
+  overallPaid: number,
+  today: Date,
+  taxYear: number,
+  payment_calendar_threshold: number,
+  dueDate: Dayjs | null,
+  installments: { order: string | null; amount: number }[],
+  variableSymbol: string,
+): Omit<ResponseInstallmentPaymentDetailDto, 'activeInstallment'> & {
+  activeInstallment?: ReplaceQrCodeWithGeneratorDto<ResponseActiveInstallmentDto>
+} => {
+  // Midnight start of the next day
+  const secondPaymentDueDatetime = dayjs
+    .tz(new Date(`${taxYear}-08-31`), bratislavaTimeZone)
+    .startOf('day')
+    .add(1, 'day')
+
+  if (overallAmount - overallPaid <= 0) {
+    return {
+      isPossible: false,
+      reasonNotPossible: InstallmentPaymentReasonNotPossibleEnum.ALREADY_PAID,
+    }
+  }
+
+  if (dayjs(today) > secondPaymentDueDatetime) {
+    return {
+      isPossible: false,
+      reasonNotPossible: InstallmentPaymentReasonNotPossibleEnum.AFTER_DUE_DATE,
+    }
+  }
+
+  if (overallAmount < payment_calendar_threshold) {
+    return {
+      isPossible: false,
+      reasonNotPossible:
+        InstallmentPaymentReasonNotPossibleEnum.BELOW_THRESHOLD,
+    }
+  }
+
+  const installmentAmounts = calculateInstallmentAmounts(
+    installments,
+    overallPaid,
+  )
+
+  const installmentDetails: ResponseInstallmentItemDto[] =
+    !dueDate ||
+    installmentAmounts[0].status === InstallmentPaidStatusEnum.PAID ||
+    installmentAmounts[0].status === InstallmentPaidStatusEnum.OVER_PAID ||
+    dueDate > dayjs(today)
+      ? [
+          {
+            installmentNumber: 1,
+            dueDate: dueDate?.toDate(),
+            status: installmentAmounts[0].status,
+            remainingAmount: installmentAmounts[0].toPay,
+          },
+          {
+            installmentNumber: 2,
+            dueDate: dayjs
+              .tz(`${taxYear}-${secondPaymentDueDate}`, bratislavaTimeZone)
+              .toDate(),
+            status: installmentAmounts[1].status,
+            remainingAmount: installmentAmounts[1].toPay,
+          },
+          {
+            installmentNumber: 3,
+            dueDate: dayjs
+              .tz(`${taxYear}-${thirdPaymentDueDate}`, bratislavaTimeZone)
+              .toDate(),
```

Comment:
if you make a datetime variable for this at the same place where `secondPaymentDueDatetime` is created, it will make it much more readable